sphinx/source/docs/user_guide/data.rst

```diff
@@ -25,69 +30,122 @@ to the ``circle`` plotting method (see :ref:`userguide_plotting` for more exampl
     p = figure()
     p.circle(x=x_values, y=y_values)
 
-When you pass in data like this, Bokeh works behind the scenes to make a
-|ColumnDataSource| for you. But learning to create and use the |ColumnDataSource|
-will enable you to access more advanced capabilities, such as streaming data,
-sharing data between plots, and filtering data.
+.. _userguide_data_numpy:
 
-ColumnDataSource
-----------------
+Providing NumPy data
+--------------------
 
-The |ColumnDataSource| is the core of most Bokeh plots, providing the data
-that is visualized by the glyphs of the plot. With the |ColumnDataSource|,
-it is easy to share data between multiple plots and widgets, such as the
-|DataTable|. When the same |ColumnDataSource| is used to drive multiple
-renderers, selections of the data source are also shared. Thus, it is possible
-to use a select tool to choose data points from one plot and have them automatically
-highlighted in a second plot (:ref:`userguide_data_linked_selection`).
-
-At the most basic level, a |ColumnDataSource| is simply a mapping between column
-names and lists of data. The |ColumnDataSource| takes a ``data`` parameter which is a dict,
-with string column names as keys and lists (or arrays) of data values as values. If one positional
-argument is passed to the |ColumnDataSource| initializer, it will be taken as ``data``. Once the
-|ColumnDataSource| has been created, it can be passed into the ``source`` parameter of
-plotting methods which allows you to pass a column's name as a stand-in for the data values:
+Similarly to using Python lists and arrays, you can also work with NumPy data
+structures in Bokeh:
+
+.. code-block:: python
+
+    import numpy as np
+    from bokeh.plotting import figure
+
+    x = [1, 2, 3, 4, 5]
+    random = np.random.standard_normal(5)
+    cosine = np.cos(x)
+
+    p = figure()
+    p.circle(x=x, y=random)
+    p.line(x=x, y=cosine)
+
+.. _userguide_data_cds:
+
+Providing data as a ColumnDataSource
+------------------------------------
+
+The |ColumnDataSource| (CDS) is the core of most Bokeh plots. It provides the
+data to the glyphs of your plot.
+
+When you pass sequences like Python lists or NumPy arrays to a Bokeh renderer,
+Bokeh automatically converts those data structures into ColumnDataSources.
+Directly working with ColumnDataSources gives you access to more advanced
+options, compared to using standard data sequences like lists and arrays.
+
+For example: Using a |ColumnDataSource| allows you to share data between
+multiple plots and widgets. If you use a single ColumnDataSource together with
+multiple renderers, those renderers also share information about data you
+select with a select tool from Bokeh's toolbar (see
+:ref:`userguide_data_linked_selection`).
+
+Think of a ColumnDataSource as a collection of sequences of data that each have
+their own, unique column name.
+
+To create a |ColumnDataSource| object, you need a Python dictionary to pass to
+the object's ``data`` parameter:
```

Comment:
Maybe: To create a *basic* CDS? Because you can also pass dataframes and groupbys and they will be adapted automatically.